provider/github/watcher.go

```diff
@@ -83,21 +94,27 @@ func (w *Watcher) watchPrs(ctx context.Context, c *Client, repos []*lookout.Repo
 				errCh <- err
 				return
 			}
-
-			select {
-			case <-ctx.Done():
-				return
-			case <-time.After(w.newInterval(c.Rate(coreCategory))):
-				continue
-			}
 		}
 	}
 }
 
 func (w *Watcher) watchEvents(ctx context.Context, c *Client, repos []*lookout.RepositoryInfo, cb lookout.EventHandler, errCh chan error) {
+	interval := minInterval
+
 	for {
 		for _, repo := range repos {
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(interval):
```

Comment:
why do we need to wait for the interval before doing anything?

Reply:
To ensure we wait in between iterations after the new `continue` if the error is `ErrGitHubAPI`.

The other options I thought of are `if/else`, which goes a bit against the early return style and avoiding nested code:

```go
func (w *Watcher) watchEvents(ctx context.Context, c *Client, repos []*lookout.RepositoryInfo, cb lookout.EventHandler, errCh chan error) {
	for {
		for _, repo := range repos {
			resp, events, err := w.doEventRequest(ctx, repo.Username, repo.Name)

			var pollInterval time.Duration

			if err != nil {
				if ErrGitHubAPI.Is(err) {
					log.With(log.Fields{"repository": repo.FullName, "response": resp}).Errorf(err, "request for events list failed")
				} else if !NoErrNotModified.Is(err) {
					errCh <- err
					return
				}
			} else {
				if err := w.handleEvents(cb, repo, resp, events); err != nil {
					errCh <- err
					return
				}

				interval := w.newInterval(c.Rate(coreCategory))
				pollInterval = c.PollInterval(eventsCategory)
				if pollInterval > interval {
					interval = pollInterval
				}
			}

			select {
			case <-ctx.Done():
				return
			case <-time.After(pollInterval):
				continue
			}
		}
	}
}
```
Or wrapping the block in a function that can return at any moment, and still end in the main last `select`:

```go
func (w *Watcher) watchEventsC(ctx context.Context, c *Client, repos []*lookout.RepositoryInfo, cb lookout.EventHandler, errCh chan error) {
	for {
		for _, repo := range repos {
			var pollInterval time.Duration

			err := func() error {
				resp, events, err := w.doEventRequest(ctx, repo.Username, repo.Name)
				if ErrGitHubAPI.Is(err) {
					log.With(log.Fields{"repository": repo.FullName, "response": resp}).Errorf(err, "request for events list failed")
					return nil
				}

				if err != nil && !NoErrNotModified.Is(err) {
					return err
				}

				if err := w.handleEvents(cb, repo, resp, events); err != nil {
					return err
				}

				interval := w.newInterval(c.Rate(coreCategory))
				pollInterval = c.PollInterval(eventsCategory)
				if pollInterval > interval {
					interval = pollInterval
				}

				return nil
			}()

			if err != nil {
				errCh <- err
				return
			}

			select {
			case <-ctx.Done():
				return
			case <-time.After(pollInterval):
				continue
			}
		}
	}
}
```

Do you think any of these 2 are better options?

Reply:
Yes. I think the second option is better. Because waiting for the interval on start instead of start watching immediately looks strange.

If you use the second option you can also move for-for loop, select and error handling to separate function and reuse it in watchPRs too.

Reply:
Pushed the changes in d6c77ad, please take a look.